File: src/api/services/netWorthService.ts
/**
 * Net Worth API Service
 * 
 * Handles net worth calculations via the backend API.
 */

import { ApiClient } from '../client';
import {
  Asset,
  Liability,
  AssetCategory,
  LiabilityCategory,
  NetWorthSummary,
  CategoryBreakdown,
} from '../../shared/types';
import { AssetApiService } from './assetService';
import { LiabilityApiService } from './liabilityService';

interface NetWorthResponse {
  total_assets: number;
  total_liabilities: number;
  net_worth: number;
  assets_by_category: Record<string, number>;
  liabilities_by_category: Record<string, number>;
  asset_breakdown: Array<{
    category: string;
    label: string;
    value: number;
    percentage: number;
    color: string;
  }>;
  liability_breakdown: Array<{
    category: string;
    label: string;
    value: number;
    percentage: number;
    color: string;
  }>;
  last_updated: string;
}

class NetWorthApiServiceClass {
  /**
   * Calculate complete net worth summary.
   */
  async calculate(): Promise<NetWorthSummary> {
    const response = await ApiClient.get<NetWorthResponse>('/net-worth');
    
    return {
      totalAssets: response.total_assets,
      totalLiabilities: response.total_liabilities,
      netWorth: response.net_worth,
      assetsByCategory: response.assets_by_category as Record<AssetCategory, number>,
      liabilitiesByCategory: response.liabilities_by_category as Record<LiabilityCategory, number>,
      lastUpdated: response.last_updated,
    };
  }

  /**
   * Get asset breakdown for charts.
   */
  async getAssetBreakdown(): Promise<CategoryBreakdown[]> {
    const response = await ApiClient.get<NetWorthResponse>('/net-worth');
    return response.asset_breakdown.map((item) => ({
      category: item.category as AssetCategory,
      label: item.label,
      value: item.value,
      percentage: item.percentage,
      color: item.color,
    }));
  }

  /**
   * Get liability breakdown for charts.
   */
  async getLiabilityBreakdown(): Promise<CategoryBreakdown[]> {
    const response = await ApiClient.get<NetWorthResponse>('/net-worth');
    return response.liability_breakdown.map((item) => ({
      category: item.category as LiabilityCategory,
      label: item.label,
      value: item.value,
      percentage: item.percentage,
      color: item.color,
    }));
  }

  /**
   * Get all assets.
   */
  async getAssets(): Promise<Asset[]> {
    return AssetApiService.list();
  }

  /**
   * Get all liabilities.
   */
  async getLiabilities(): Promise<Liability[]> {
    return LiabilityApiService.list();
  }

  /**
   * Add a new asset.
   */
  async addAsset(data: Omit<Asset, 'id' | 'createdAt' | 'updatedAt'>): Promise<Asset> {
    return AssetApiService.create(data);
  }

  /**
   * Update an existing asset.
   */
  async updateAsset(
    id: string,
    data: Partial<Omit<Asset, 'id' | 'createdAt' | 'updatedAt'>>
  ): Promise<Asset | null> {
    return AssetApiService.update(id, data);
  }

  /**
   * Delete an asset.
   */
  async deleteAsset(id: string): Promise<boolean> {
    return AssetApiService.delete(id);
  }

  /**
   * Add a new liability.
   */
  async addLiability(
    data: Omit<Liability, 'id' | 'createdAt' | 'updatedAt'>
  ): Promise<Liability> {
    return LiabilityApiService.create({
      category: data.category,
      name: data.name,
      balance: data.balance,
      interest_rate: data.interestRate,
    });
  }

  /**
   * Update an existing liability.
   */
  async updateLiability(
    id: string,
    data: Partial<Omit<Liability, 'id' | 'createdAt' | 'updatedAt'>>
  ): Promise<Liability | null> {
    return LiabilityApiService.update(id, {
      category: data.category,
      name: data.name,
      balance: data.balance,
      interest_rate: data.interestRate,
    });
  }

  /**
   * Delete a liability.
   */
  async deleteLiability(id: string): Promise<boolean> {
    return LiabilityApiService.delete(id);
  }
}

// Export singleton instance
export const NetWorthApiService = new NetWorthApiServiceClass();

File: backend/app/services/net_worth_service.py
"""
Net Worth Service

Business logic for net worth calculations.
"""

from datetime import datetime
from typing import List
from sqlalchemy.orm import Session

from app.repositories.asset_repository import asset_repository
from app.repositories.liability_repository import liability_repository
from app.schemas.net_worth import NetWorthSummary, CategoryBreakdown
from app.schemas.asset import ASSET_CATEGORY_INFO, AssetCategory
from app.schemas.liability import LIABILITY_CATEGORY_INFO, LiabilityCategory


class NetWorthService:
    """Service for net worth calculations."""
    
    def calculate(self, db: Session) -> NetWorthSummary:
        """Calculate complete net worth summary."""
        total_assets = asset_repository.get_total_value(db)
        total_liabilities = liability_repository.get_total_balance(db)
        net_worth = total_assets - total_liabilities
        
        assets_by_category = asset_repository.get_value_by_category(db)
        liabilities_by_category = liability_repository.get_balance_by_category(db)
        
        asset_breakdown = self._get_asset_breakdown(assets_by_category, total_assets)
        liability_breakdown = self._get_liability_breakdown(liabilities_by_category, total_liabilities)
        
        return NetWorthSummary(
            total_assets=total_assets,
            total_liabilities=total_liabilities,
            net_worth=net_worth,
            assets_by_category=assets_by_category,
            liabilities_by_category=liabilities_by_category,
            asset_breakdown=asset_breakdown,
            liability_breakdown=liability_breakdown,
            last_updated=datetime.utcnow(),
        )
    
    def _get_asset_breakdown(
        self, 
        by_category: dict, 
        total: float
    ) -> List[CategoryBreakdown]:
        """Get asset breakdown for charts."""
        breakdown = []
        
        for category, value in by_category.items():
            if value > 0:
                try:
                    cat_enum = AssetCategory(category)
                    info = ASSET_CATEGORY_INFO.get(cat_enum, {})
                except ValueError:
                    info = {"label": category, "color": "#a0a0a0"}
                
                percentage = (value / total * 100) if total > 0 else 0
                
                breakdown.append(CategoryBreakdown(
                    category=category,
                    label=info.get("label", category),
                    value=value,
                    percentage=round(percentage, 1),
                    color=info.get("color", "#a0a0a0"),
                ))
        
        # Sort by value descending
        breakdown.sort(key=lambda x: x.value, reverse=True)
        return breakdown
    
    def _get_liability_breakdown(
        self, 
        by_category: dict, 
        total: float
    ) -> List[CategoryBreakdown]:
        """Get liability breakdown for charts."""
        breakdown = []
        
        for category, value in by_category.items():
            if value > 0:
                try:
                    cat_enum = LiabilityCategory(category)
                    info = LIABILITY_CATEGORY_INFO.get(cat_enum, {})
                except ValueError:
                    info = {"label": category, "color": "#c7b8b8"}
                
                percentage = (value / total * 100) if total > 0 else 0
                
                breakdown.append(CategoryBreakdown(
                    category=category,
                    label=info.get("label", category),
                    value=value,
                    percentage=round(percentage, 1),
                    color=info.get("color", "#c7b8b8"),
                ))
        
        # Sort by value descending
        breakdown.sort(key=lambda x: x.value, reverse=True)
        return breakdown


# Singleton instance
net_worth_service = NetWorthService()

